### PR TITLE
[FLOC-3498] Use ClusterHQ fork of testtools

### PR DESCRIPTION
**Pull request by jml**
It's taking days to get patches landed in testtools and these delays are having a direct & significant impact on our development velocity (flaky tests; debugging info from failing tests). This patch circumvents  that by using a temporary, dedicated fork: ClusterHQ/testtools#clusterhq-fork

The trade-off is that we'll probably have to do some rework in exchange for advancing our learning from the future to today.

The fork currently incorporates the latest versions of testing-cabal/testtools#165 and testing-cabal/testtool#171.

The astute will note that we've _already_ learned something, which is that testing-cabal/testtools#165 isn't enough for a `RunTest` implementation to be able to run a test case twice. I plan to address that in a follow-up patch on testtools, since 165 is already huge.

### CODE REVIEW

File: flocker/testtools/_flaky.py [hunk 1]
```diff
@@ -238,8 +238,9 @@ def _attempt_test(self, case):
             result was and ``details`` is a dictionary of testtools details.
         """
         tmp_result = testtools.TestResult()
-        # XXX: Work around https://bugs.launchpad.net/testtools/+bug/1517879
-        _reset_case(case)
+        # XXX: jml: Argh! Of course putting _reset in TestCase.run() doesn't
```
Comment (sarum90) [inline flocker/testtools/_flaky.py]: I don't understand this XXX.

File: setup.py [hunk 2]
```diff
@@ -108,6 +108,12 @@
 
     cmdclass=versioneer.get_cmdclass(),
 
+    dependency_links = [
+        # Use our fork of testtools until #165, #171, and #172 are merged and
+        # released. See FLOC-3498.
+        "git+git://github.com/ClusterHQ/testtools@clusterhq-fork#egg=testtools-1.8.2chq1",  # noqa
```
Comment (sarum90) [inline setup.py]: is git+git a thing? It's a hard thing to google search for so I couldn't verify.

**Comment by sarum90** [comment]
This looks fine to me. As far as I'm concerned if this makes your work go faster and it doesn't break any clusterHQ developers dev environment or the jenkins tests it should be submitted. That said... it seems to still make Jenkins be completely red, so I'm rejecting only because I'm unsure how complex the fix to that will be.

Also I wanted to double check:
-Does this work on the Vagrant Dev box?
-Does this work on an OSX dev environment?

^^ I'm willing to test one or both of those for you as part of the review.

I did not attempt to understand every part of this PR, instead basically checking it for typos and the like. It looks reasonable to me, and I'm not worried about running our own fork of a testing library.

File: flocker/testtools/_flaky.py (hunk 1 above)
Comment (jml) [inline flocker/testtools/_flaky.py]: Fixed.

File: setup.py (hunk 2 above)
Comment (jml) [inline setup.py]: Yes, it is. Added a comment explaining what's going on & linked to the docs.

**Comment by jml** [comment]
* Fixed Jenkins being completely red (now completely green, although not yet seen results of last push, which only changed comments)
* Confirm it works on OS X dev environment (that's what I used to develop)
* No idea whether it works on vagrant dev box—we're only half supporting it now—but feel free to test and I'll do any shallow fixes required.
* I've updated the bug to have more explanation of why we're forking.

Thanks!

**Comment by sarum90** [comment]
Working on OSX and in the automated tests is good enough for me.

So long as this isn't breaking peoples builds go ahead and merge! :)

**Pull request merged**